File: 2015 letni (Spring) - 2018 zimowy (Fall)/PP-Fall-2018-2019/old/Stationary/2018-10-10/Heron.java
class Heron {

  static double square(double x) {
    return x * x;
  }

  static double abs(double x) {
    if (x < 0) {
      return -x;
    } else {
      return x;
    }
  }

  static boolean isGoodEnough(double y, double x) {
    return abs(square(y) - x) <= 0.0001;
  }

  static double avg(double x, double y) {
    return (x + y) / 2.0;
  }

  static double enhance(double y, double x) {
    return avg(y, x / y);
  }

  static double sqrt(double y, double x) {
    if (isGoodEnough(y, x)) {
      return y;
    } else {
      return sqrt(enhance(y, x), x);
    }
  }

  public static void main(String[] args) {
    System.out.println(sqrt(1, 2));
  }

}

File: 2021 letni/FP (niestacjonarne)/fp1/src/fp1/heron.clj
(ns fp1.heron)

(defn abs
  [x]
  (if (< x 0)
    (- x)
    x))

(defn square
  [x]
  (* x x))

(defn good-enough?
  [G x]
  (<= (abs (- x (square G))) 0.00000001))

(defn avg
  [x y]
  (/ (+ x y) 2))

(defn improve
  [G x]
  (avg G (/ x G)))

(defn heron-sqrt
  [G x]
  (println "G is" G)
  (if (good-enough? G x)
    G

    ;; otherwise,
    (heron-sqrt (improve G x) x)))

#_(time (heron-sqrt 1 2))

#_(-
    (double (heron-sqrt 1 2))
    (Math/sqrt 2))

;; Definicja procedury
#_(defn <procedure-name>
    [<arg-1> <arg-2> ... <arg-n>]
    <body>)

;; Wyrażenie warunkowe
#_(if <predicate>
    <consequent>
    <alternative>)

;; JS ternary-conditional expression
;; <predicate> ? <consequent> : <alternative>;
;; np. let v1 = s1 === 'aaa' ? 2 : 3;

;; Zadanie 1. Proszę uzależnić ilość kroków wykonywanych przez procedurę heron-sqrt
;;            od pewnego parametru tej procedury określającego dokładność
;;            (zamiast stałej 0.00000001).

;; Zadanie 2. Proszę uzależnić ilość kroków wykonywanych przez procedurę heron-sqrt
;;            od pewnego parametru (n - liczby całkowitej) oznaczającego tę ilość.

;; Zadanie 3. Proszę zrealizować procedurę liczącą pierwiastek sześcienny podanej liczby x.
;;            Hint: SICP Exercise 1.8
